manim/mobject/three_dimensions.py

```diff
@@ -189,3 +200,426 @@ def generate_points(self):
         Cube.generate_points(self)
         for dim, value in enumerate(self.dimensions):
             self.rescale_to_fit(value, dim, stretch=True)
+
+
+class Cone(ParametricSurface):
+    """A circular cone.
+    Can be defined using 2 parameters: its height, and its base radius.
+    The polar angle, theta, can be calculated using arctan(base_radius /
+    height) The spherical radius, r, is calculated using the pythagorean
+    theorem.
+
+    Examples
+    --------
+    .. manim:: ExampleCone
+        :save_last_frame:
+
+        class ExampleCone(ThreeDScene):
+            def construct(self):
+                axes = ThreeDAxes()
+                cone = Cone(direction=X_AXIS+Y_AXIS+2*Z_AXIS)
+                self.set_camera_orientation(phi=5*PI/11, theta=PI/9)
+                self.add(axes, cone)
+
+    Parameters
+    --------
+    base_radius : :class:`float`
+        The base radius from which the cone tapers.
+    height : :class:`int`
+        The height measured from the plane formed by the base_radius to the apex of the cone.
+    direction :  :class:`np.array`
+        The direction of the apex.
+    show_base : :class:`bool`
+        Whether to show the base plane or not.
+    v_min : :class:`float`
+        The azimuthal angle to start at.
+    v_max : :class:`float`
+        The azimuthal angle to end at.
+    u_min : :class:`float`
+        The radius at the apex.
+    checkerboard_colors : :class:`bool`
+        Show checkerboard grid texture on the cone.
+    """
+
+    def __init__(
+        self,
+        base_radius=1,
+        height=1,
+        direction=Z_AXIS,
+        show_base=False,
+        v_min=0,
+        v_max=TAU,
+        u_min=0,
+        checkerboard_colors=False,
+        **kwargs
+    ):
+        self.base_radius = base_radius
+        self._height = height
+        self.show_base = show_base
+        self.direction = direction
+        ParametricSurface.__init__(
+            self,
+            self.func,
+            v_min=v_min,
+            v_max=v_max,
+            u_min=u_min,
+            u_max=np.sqrt(base_radius ** 2 + height ** 2),
+            checkerboard_colors=checkerboard_colors,
+            **kwargs
+        )
+        # used for rotations
+        self._current_theta = 0
+        self._current_phi = 0
+
+        if self.show_base:
+            self.base_circle = Dot(
+                point=self._height * IN,
+                radius=self.base_radius,
+                color=self.fill_color,
+                fill_opacity=self.fill_opacity,
+            )
+            self.add(self.base_circle)
+
+        self._rotate_to_direction()
+
+    @property
+    def theta(self):
+        return PI - np.arctan(self.base_radius / self._height)
```

Comment:
Does `self.base_radius` or `self._height` ever change without a user setting them? Because otherwise this doesn't need to be a property and can just be calculated on init, which would also make it so `self._height` doesn't need to be set at all, which I think would be an improvement.

Reply:
Removed.